licensing/src/main/java/io/crate/license/LicenseService.java

```diff
@@ -129,6 +122,81 @@ public void onFailure(Exception e) {
         }
     }
 
+    //@VisibleForTesting
+    boolean verifyLicense(LicenseKey licenseKey) {
+        try {
+            DecodedLicense decodedLicense = decodeLicense(licenseKey);
+            DecryptedLicenseData licenseInfo = decryptLicenseContent(decodedLicense.encryptedContent());
+
+            if (licenseInfo.isExpired()) {
+                //LOGGER.warn("Unable to validate [{}] as it is expired", license);
+                return false;
+            }
+            if (decodedLicense.type() == LicenseType.ENTERPRISE) {
+                return verifySignature(licenseInfo);
+            }
+            return true;
+        } catch (IOException e) {
+            return false;
+        }
+    }
+
+    //@VisibleForTesting
+    static DecryptedLicenseData decryptLicenseContent(byte[] encryptedContent) throws IOException {
+        byte[] decryptedContent = Cryptos.decrypt(encryptedContent);
+        return DecryptedLicenseData.fromFormattedLicenseData(decryptedContent);
+    }
+
+
+    //@VisibleForTesting
+    static byte[] encryptLicenseContent(byte[] content) {
+        return Cryptos.encrypt(content);
+    }
+
+    private static boolean verifySignature(final DecryptedLicenseData license) {
+        final byte[] publicKeyBytes;
+        try (InputStream is = LicenseService.class.getResourceAsStream("/public.key")) {
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
+            Streams.copy(is, out);
+            publicKeyBytes = out.toByteArray();
+        } catch (IOException ex) {
+            throw new IllegalStateException(ex);
+        }
+        return verifySignature(license, publicKeyBytes);
+    }
+
+    private static boolean verifySignature(final DecryptedLicenseData licenseData, byte[] publicKeyBytes) {
+        try {
+            // init signature with public key
+            Signature rsa = Signature.getInstance("SHA512withRSA");
```

Comment:
Would it be possible to split the crypto code more from the licensing specific code?

We're creating loose contracts here in that the used algorithms have to match. E.g. `Cryptos` has a `RSA` constant and here we're defining `SHA512WithRSA` but there is no gurantee that the `readPublicKey` call returns a public key that "fits"

Reply:
done